utility/utility.go

```diff
@@ -1,111 +1,283 @@
+	if fm == nil {
+		return nil
+	}
+	session.Save(r, w)
+	return fm
+}
+
+func RenderTemplate(w http.ResponseWriter, r *http.Request, template string, data interface{}) {
+	session, _ := Store.Get(r, os.Getenv("SESSION_NAME"))
+	tmplData := make(map[string]interface{})
+	tmplData["data"] = data
+	tmplData["flash"] = viewFlash(w, r)
+	tmplData["session"] = session.Values["email"]
+	View.ExecuteTemplate(w, template, tmplData)
+}
+
+func ParseDataFromPostFormRequest(r *http.Request) (map[string]interface{}, error) {
+	formData := make(map[string]interface{})
+	err := r.ParseForm()
+	if err != nil {
+		return formData, err
+	}
+
+	// Iterate through the form values
+	for key, values := range r.Form {
+		// If there's only one value for the key, store it directly
+		if len(values) == 1 {
+			formData[key] = values[0]
+		} else {
+			// If there are multiple values, store them as a slice
+			formData[key] = values
+		}
+	}
+
+	return formData, nil
 }
 
-func viewFlash(w http.ResponseWriter, r *http.Request) interface{}{
-  session, err := Store.Get(r, os.Getenv("SESSION_NAME"))
-  if err != nil {
-    http.Error(w, err.Error(), http.StatusInternalServerError)
-  }
-  fm := session.Flashes("message")
-  if fm == nil {
-    return nil
-  }
-  session.Save(r, w)
-  return fm
+func ParseDataFromJson(r *http.Request) (map[string]interface{}, error) {
+	var jsonDataMap map[string]interface{}
+	result := make(map[string]interface{})
+
+	// Read JSON data from the HTTP request body
+	decoder := json.NewDecoder(r.Body)
+	err := decoder.Decode(&jsonDataMap)
+	if err != nil {
+		return result, err
+	}
+
+	// Close the request body to prevent resource leaks
+	defer r.Body.Close()
+	// Iterate through the JSON values
+	for key, value := range jsonDataMap {
+		result[key] = value
+	}
+
+	return result, nil
+}
+
+func StrictParseDataFromJson(r *http.Request, structure interface{}) error {
+	err := json.NewDecoder(r.Body).Decode(structure)
+	if err != nil {
+		return err
+	}
+	return err
+
+}
+
+func StrictParseDataFromPostFormRequest(r *http.Request, structure interface{}) error {
+	err := r.ParseForm()
+	if err != nil {
+		return err
+	}
+	// Use reflection to set field values based on form data
+	structValue := reflect.ValueOf(structure)
+	if structValue.Kind() != reflect.Ptr || structValue.Elem().Kind() != reflect.Struct {
+		return errors.New("invalid structure, must be a pointer to a struct")
+	}
+	structElem := structValue.Elem()
+	for key, values := range r.Form {
+		field := structElem.FieldByName(key)
+		if !field.IsValid() {
+			// Skip fields that don't exist in the structure
+			continue
+		}
+		// Handle fields with different types (e.g., slice or single value)
+		if len(values) == 1 {
+			value := values[0]
+			//conversion of the data came according to the feilds values that are used in the struct
+			switch field.Kind() {
+			case reflect.Int:
+				intValue, err := strconv.Atoi(value)
+				if err != nil {
+					return err
```

Comment:
Not sure what these errros would look like but would it give any indication to the user of where it happened ?, like which key was the culprit behind the error ?